### PR TITLE
Keypaths

### PR DESCRIPTION
**Pull request by jckarter**
Merging WIP of SE-0161 into master.

### CODE REVIEW

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test

File: lib/SIL/SILVerifier.cpp [hunk 1]
```diff
@@ -3665,6 +3665,90 @@ class SILVerifier : public SILVerifierBase<SILVerifier> {
...
+        case KeyPathPatternComponent::Kind::SettableProperty: {
+          require(component.getComputedPropertyIndices().empty(),
+                  "subscripts not implemented");
+
+          // TODO: Verify the signatures of the getter and setter
+          break;
+        }
+        }
+        
+        baseTy = componentTy;
```
Comment (slavapestov) [inline lib/SIL/SILVerifier.cpp]: Is componentTy ever dependent? Use after free

Comment (jckarter) [inline lib/SIL/SILVerifier.cpp]: It should be substituted into the SILFunction context.

Comment (slavapestov) [inline lib/SIL/SILVerifier.cpp]: What do you need the generic context for then? The issue is that after a context is popped, any dependent types vended by type lowering are now invalid.

File: lib/SILGen/SILGenExpr.cpp [hunk 2]
```diff
@@ -2383,8 +2411,443 @@ RValue RValueEmitter::visitObjCSelectorExpr(ObjCSelectorExpr *e, SGFContext C) {
...
+    ->getCanonicalType();
+  auto genericEnv = typeContext->getGenericEnvironmentOfContext();
+  auto genericSig = typeContext->getGenericSignatureOfContext()
+    ? typeContext->getGenericSignatureOfContext()->getCanonicalSignature()
+    : nullptr;
+  
+  SILType loweredBaseTy, loweredPropTy;
+  {
+    GenericContextScope scope(SGF.SGM.Types, genericSig);
+    loweredBaseTy = SGF.getLoweredType(AbstractionPattern::getOpaque(),
```
Comment (slavapestov) [inline lib/SILGen/SILGenExpr.cpp]: This is definitely wrong -- loweredBaseTy outlives the scope

File: stdlib/public/core/KeyPath.swift [hunk 3]
```diff
@@ -0,0 +1,1739 @@
...
+
+// FIXME(ABI): The type relationships between KeyPath append operands are tricky
+// and don't interact well with our overriding rules. Hack things by injecting
+// a bunch of `appending` overloads as protocol extensions so they aren't
+// constrained by being overrides, and so that we can use exact-type constraints
+// on `Self` to prevent dynamically-typed methods from being inherited by
+// statically-typed key paths.
+public protocol _AppendKeyPath {}
+
+extension _AppendKeyPath where Self == AnyKeyPath {
```
Comment (slavapestov) [inline stdlib/public/core/KeyPath.swift]: Naughty naughty! This doesn't really work properly. If you need it to work, there are a couple of bugs we need to fix.

File: stdlib/public/core/KeyPath.swift [hunk 4]
```diff
@@ -0,0 +1,1739 @@
...
+    path: KeyPath<Value, AppendedValue>
+  ) -> KeyPath<Root, AppendedValue>
+  where Self: KeyPath<Root, Value> {
+    return _appendingKeyPaths(root: self, leaf: path)
+  }
+
+  /* TODO
+  public func appending<Root, Value, Leaf>(
+    path: Leaf,
+    // FIXME: Satisfy "Value generic param not used in signature" constraint
```
Comment (slavapestov) [inline stdlib/public/core/KeyPath.swift]: Looks like a bug in the 'all parameters are used' check -- the signature looks fine. Can you file a bug?

File: stdlib/public/core/KeyPath.swift [hunk 5]
```diff
@@ -0,0 +1,1739 @@
...
+extension _AppendKeyPath /* where Self == ReferenceWritableKeyPath<T,U> */ {
+  public func appending<Root, Value, AppendedValue>(
+    path: WritableKeyPath<Value, AppendedValue>
+  ) -> ReferenceWritableKeyPath<Root, AppendedValue>
+  where Self == ReferenceWritableKeyPath<Root, Value> {
+    return _appendingKeyPaths(root: self, leaf: path)
+  }
+}
+
+// internal-with-availability
```
Comment (slavapestov) [inline stdlib/public/core/KeyPath.swift]: You can write @_versioned now

File: stdlib/public/core/KeyPath.swift [hunk 6]
```diff
@@ -0,0 +1,1739 @@
...
+  where Self == WritableKeyPath<Root, Value> {
+    return _appendingKeyPaths(root: self, leaf: path)
+  }
+}
+
+extension _AppendKeyPath /* where Self == ReferenceWritableKeyPath<T,U> */ {
+  public func appending<Root, Value, AppendedValue>(
+    path: WritableKeyPath<Value, AppendedValue>
+  ) -> ReferenceWritableKeyPath<Root, AppendedValue>
+  where Self == ReferenceWritableKeyPath<Root, Value> {
```
Comment (slavapestov) [inline stdlib/public/core/KeyPath.swift]: You're using my other naughty trick too :)

**Comment by slavapestov** [comment]
Looks like you found a couple of nice generics and optimizer bugs too, sweet.

File: stdlib/public/core/KeyPath.swift (hunk 4 above)
Comment (slavapestov) [inline stdlib/public/core/KeyPath.swift]: Maybe it's already fixed?

```
(swift) class C<T, U> {}
(swift) func f<T, U, V>() -> T where T : C<U, V> {}
<REPL Input>:1:43: error: missing return in a function expected to return 'T'
func f<T, U, V>() -> T where T : C<U, V> {}
                                          ^
```

File: lib/SIL/SILVerifier.cpp (hunk 1 above)
Comment (jckarter) [inline lib/SIL/SILVerifier.cpp]: Good point, maybe I don't need it now. I had gone back and forth here a bit.

File: stdlib/public/core/KeyPath.swift (hunk 3 above)
Comment (jckarter) [inline stdlib/public/core/KeyPath.swift]: What exactly doesn't work? It seems to do what I'm abusing it for, making the methods only available exactly on AnyKeyPath.

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by swiftix** [comment]
@jckarter The AST printer fix should be probably a separate PR as it is not specific to KeyPaths.

**Comment by jckarter** [comment]
It's blocking me from landing keypaths.

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@ddunbar @rballard This swiftpm test failure look familiar to you?
https://ci.swift.org/job/swift-PR-Linux/6945/consoleFull#1552161685fca400bf-2f4a-462e-b517-e058d770b2d7

**Comment by jckarter** [comment]
@akyrtzi @benlangmuir @nkcsgexi Is this just nondeterminism in this SourceKit test?

https://ci.swift.org/job/swift-PR-osx/6963/consoleFull#-1727017123ee1a197b-acac-4b17-83cf-a53b95139a76

**Comment by jckarter** [comment]
@swift-ci Please smoke test

**Comment by ddunbar** [comment]
The swiftpm failure is a known Swift 4 incompatibility -- @aciidb0mb3r did we slip in a change to Swift 4 too early?

**Comment by aciidb0mb3r** [comment]
Most likely fallout from https://github.com/apple/swift-package-manager/pull/1118, stay tuned for fix.

**Comment by aciidb0mb3r** [comment]
This will fix it: https://github.com/apple/swift-package-manager/pull/1122

**Comment by jckarter** [comment]
@swift-ci Please smoke test

**Comment by jckarter** [comment]
@swift-ci Please smoke test OS X

**Comment by benlangmuir** [comment]
RE: the SourceKit test.  I'll fix this to stop checking USRs.  I attempted to do that in #8856, but apparently I missed two cases.

**Comment by jckarter** [comment]
@swift-ci Please smoke test

**Comment by jckarter** [comment]
Thanks @benlangmuir! I added a commit to disable the test in 73383b7, do you want me to wait for your fix?

**Comment by benlangmuir** [comment]
If you don't mind waiting I just kicked off testing on https://github.com/apple/swift/pull/8929.  It's not actually non-deterministic (or shouldn't be!), it's just depending on mangling of a couple of stdlib symbols.

**Comment by jckarter** [comment]
OK, thanks @benlangmuir.

**Comment by jckarter** [comment]
@swift-ci Please smoke test

**Comment by jckarter** [comment]
@swift-ci Please smoke test

**Comment by jckarter** [comment]
@swift-ci Please smoke test

**Comment by jckarter** [comment]
@swift-ci Please smoke test

**Pull request merged**